### PR TITLE
added functionality to filter results by keys in _list.py

### PR DESCRIPTION
**Pull request by pietropra**
*Description of changes:* Added the ability to filter out results by specifying a keyword for keys/prefix in wr.s3.list_objects()

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by igorborgest — CHANGES_REQUESTED** [review]
Thanks @pietropra ! It's a cool feature.

I've leaved some comments and also missed some test to cover it.

File: awswrangler/s3/_list.py [hunk 1]
```diff
@@ -117,6 +120,9 @@ def _list_objects(  # pylint: disable=too-many-branches
     if _ignore_suffix is not None:
         paths = [p for p in paths if p.endswith(tuple(_ignore_suffix)) is False]
 
+    if excluded_keys is not None:
+        paths = [p for p in paths if not any([search(key, p) for key in _excluded_keys])]
```
[the remark this answers is not in the document]
Comment (pietropra) [reply awswrangler/s3/_list.py:124]: Thanks Igor, good suggestions. I will work on this further and re-submit

[the remark this answers is not in the document]
Comment (pietropra) [reply awswrangler/s3/_list.py]: HI, I have pushed a new commit

I changed parse_path to return a list of key parts, and a check in list_objects to return filtered paths if any key part matches the excluded keys in the excluded_keys argument.

Let me know if this is any better?

**Comment by pietropra** [comment]
@igorborgest 

Now `list_objects `works by breaking down the key after bucket into parts and, if `excluded _keys` is not `None`, it filters out the objects that have a key from the excluded keys (either a String or List of Strings).

This is an example of `parse_path` behaves when `excluded_keys `is not `None`:
```
paths = [
    's3://bucket/prefix/archive',
    's3://bucket/prefix/my-foo-archive',
]

excluded_keys = [
    'archive',
    'logs',
]

paths_to_keep = []
for p in paths:
    _, prefix, keyparts = parse_path(p, multipart=True)
    if not any(key in keyparts for key in excluded_keys):
        paths_to_keep.append(p)

paths_to_keep
```
`paths_to_keep` becomes `['s3://bucket/prefix/my-foo-archive']`

Mypy is throwing some errors for how the tuple resulting from `parse_path `is unpacked in `list_objects`: the way is implemented is checking if a `multipart `argument is `True `(default `False`). 
Maybe another way to go about it it might be to separate the functions e.g. creating something like `list_and_filter_objects`, but I tried to keep the change minimal, what do you think?

**Comment by igorborgest** [comment]
Yeah, unfortunately mypy will only consider this two different possibilities of return types safe if you check it before unpack for ALL usages. But event tough I don't think this is the best way. A new parse_path function specialized to return all the parts would be better.

But before this change I would like to take a step back and understand better the final goal. The idea is to filter by black listed filenames in the end of the object key OR black listed directories in the middle of the object key?

**Comment by igorborgest** [comment]
closing by inactivity

**Pull request closed without merging**